CARDS-2725: Add option to remove @ from CSV export headers

acrowthe opened the pull request:
- Add -csvStripHeader option to csv exports that removes `@` symbol
- Add this option to prems labelled form export

Testing (prems runmode):
- From the admin view, export the inpatient survey with labels and identifiers checked. Test various options for the header label replacement and header id replacement fields. e.g.
    ```
    @=# ,        Identifier=ID
    Patient=User
    ```

acrowthe committed:
CARDS-2725: Add option to remove @ from CSV export headers
- Add -csvColumnReplace option to csv exports
  - Accepts a regex to replace and a string to replace with, eg. .csvColumnReplace:/@=#

marta- commented:
The new option should be added to the [export ui](https://github.com/data-team-uhn/cards/blob/dev/modules/data-entry/src/main/frontend/src/dataHomepage/ExportButton.jsx#L106) as well.

marta- commented:
As we are going to pull the YE code in a separate repo eventually, and we hope to import the commit history on those files from this repo to the new one, could the change to the prems feature file be done in a separate commit?

acrowthe committed:
CARDS-2725: Add option to remove @ from CSV export headers
Improve UI and split replacement into two fields, one for label and one for identifier

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/ExportButton.jsx:
```diff
@@ -333,7 +346,34 @@ function ExportButton(props) {
+            <Grid size={8}>
+              <TextField
+                variant="standard"
+                helperText="Regex based replace all. Format: <regex to find>=<value to replace with>,<regex2>=<value2>,..."
```
```suggestion
                helperText="Format: <regex to find>=<value to replace with>,<regex2>=<value2>,..."
```

marta- replied:
```suggestion
                helperText="List of pairs `<regex to find>=<value to replace with>` separated by comma or newline. Example: <regex_1>=<value_1>,<regex_2>=<value_2>,..."
```

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/ExportButton.jsx:
```diff
@@ -186,6 +189,16 @@ function ExportButton(props) {
       path += ".csvHeader:raw";
     }
+    if (csvReplaceColumnLabels) {
+      csvReplaceColumnLabels.split(",").forEach(replacement => {
```
Since we made the text field multiline, could we be a bit more flexible with the separators such as allowing "\s*,\s*" and newline?

acrowthe committed:
CARDS-2725: Add option to remove @ from CSV export headers
Improve export UI with newline delimiter for replacements and better helper text
Fix backend CSV replace when multiple replacements are present

sdumitriu commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/serialize/QuestionnaireToCsvProcessor.java:
```diff
@@ -135,6 +143,51 @@ private String processQuestionnaire(final JsonObject questionnaire, final Resour
+            }
+
+            String[] pieces = decodedPath.substring(startIndex + name.length(), endIndex)
+                .split("=", 2);
```
This doesn't really allow using `.` anywhere in the setting and `=` in the left side. I recommend doing something like what is done in [the parsing of data filters](https://github.com/data-team-uhn/cards/blob/3396a617dd60038df5ff1d235739e006bbca5355/modules/utils/src/main/java/io/uhndata/cards/serialize/internal/DefaultDataFiltersParser.java#L48), which handles `.` correctly, and extending that to support escaped `\=`.

Why is `URLDecoder.decode` needed?

sdumitriu commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/serialize/QuestionnaireToCsvProcessor.java:
```diff
@@ -135,6 +143,51 @@ private String processQuestionnaire(final JsonObject questionnaire, final Resour
+            String[] pieces = decodedPath.substring(startIndex + name.length(), endIndex)
+                .split("=", 2);
+            String left = pieces[0];
+            String right = pieces.length == 2 ? pieces[1] : pieces[0];
```
Why would we allow replacing the same thing with itself? I would discard if there's no right side.

sdumitriu replied:
Ah, I see, this is also used for the extra columns to include now, which do allow just specifying one value.

sdumitriu commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/serialize/QuestionnaireToCsvProcessor.java:
```diff
@@ -448,4 +483,26 @@ private String getStringValue(final JsonValue value, final String nodeType)
         }
     }
+
+    private class StringPair
```
I would use `org.apache.commons.lang3.tuple.Pair`, or `ImmutablePair`.

sdumitriu commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/serialize/QuestionnaireToCsvProcessor.java:
```diff
@@ -135,6 +143,51 @@ private String processQuestionnaire(final JsonObject questionnaire, final Resour
+        throws IOException
+    {
+        if (replacements.size() > 0) {
+            List<Pattern> patterns = replacements.stream().map(replacement -> Pattern.compile(replacement.getLeft()))
```
Why not compile when extracting, instead of `<String,String>` store `<Pattern,String>`.

sdumitriu committed:
CARDS-2743: Add a utility class for parsing selectors and selector options from a resolution path info string

sdumitriu committed:
CARDS-2725: Add option to remove @ from CSV export headers
Use the new SelectorUtils class for parsing the headers
